src/Analyzers/CSharp/Analyzers/ConvertProgram/ConvertProgramAnalysis_TopLevelStatements.cs

```diff
@@ -0,0 +1,159 @@
+            // `static Main` immediately.
+            if (!methodDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword) ||
+                methodDeclaration.TypeParameterList is not null ||
+                methodDeclaration.Identifier.ValueText != WellKnownMemberNames.EntryPointMethodName ||
+                methodDeclaration.Parent is not TypeDeclarationSyntax containingTypeDeclaration ||
+                methodDeclaration.Body == null)
+            {
+                return false;
+            }
+
+            // If the compilation options specified a type name that Main should be found in, then do a quick check that
+            // our containing type matches that.
+            if (mainTypeName != null && containingTypeDeclaration.Identifier.ValueText != mainTypeName)
+                return false;
+
+            // If the user renamed the 'args' parameter, we can't convert to top level statements.
+            if (methodDeclaration.ParameterList.Parameters.Count == 1 &&
+                methodDeclaration.ParameterList.Parameters[0].Identifier.ValueText != "args")
+            {
+                return false;
+            }
+
+            // Found a suitable candidate.  See if this matches the entrypoint the compiler has actually chosen.
+            var entryPointMethod = semanticModel.Compilation.GetEntryPoint(cancellationToken);
+            if (entryPointMethod == null)
+                return false;
+
+            var thisMethod = semanticModel.GetDeclaredSymbol(methodDeclaration);
+            if (!entryPointMethod.Equals(thisMethod))
+                return false;
+
+            // We found the entrypoint.  However, we can only effectively convert this to top-level-statements
+            // if the existing type is amenable to that.
+            canConvertToTopLevelStatements = TypeCanBeConverted(entryPointMethod.ContainingType, containingTypeDeclaration);
+            return true;
+        }
+
+        private static bool TypeCanBeConverted(INamedTypeSymbol containingType, TypeDeclarationSyntax typeDeclaration)
+        {
+            // Can't convert if our Program type derives or implements anything special.
+            if (containingType.BaseType?.SpecialType != SpecialType.System_Object)
+                return false;
+
+            if (containingType.AllInterfaces.Length > 0)
+                return false;
+
+            // Too complex to convert many parts to top-level statements.  Just bail on this for now.
+            if (containingType.DeclaringSyntaxReferences.Length > 1)
+                return false;
+
+            // Too complex to support converting a nested type.
+            if (containingType.ContainingType != null)
+                return false;
+
+            // If the type wasn't internal it might have been public and something outside this assembly might be using it.
+            if (containingType.DeclaredAccessibility != Accessibility.Internal)
```

Comment:
It seems the goal is to bail out for `public`, not for non-internal. The suggested change might not have any value today, but may be relevant depending on how features like file-private accessibilities will be implemented.

```suggestion
            // If the type wasn't internal it might have been public and something outside this assembly might be using it.
            if (containingType.DeclaredAccessibility == Accessibility.Public)
```

Reply:
sounds good :)